parser/src/string.rs

```diff
@@ -0,0 +1,189 @@
+use crate::ast::{Constant, Expr, ExprKind};
+use crate::error::{LexicalError, LexicalErrorType};
+
+pub fn parse_implicit_concatenation(values: Vec<Expr>) -> Result<Expr, LexicalError> {
+    Ok(if values.len() > 1 {
+        // As in CPython, use the kind of the first Expression.
+        let kind = if let ExprKind::Constant { kind, .. } = &values[0].node {
+            kind.clone()
+        } else {
+            None
+        };
+
+        // Preserve the initial location.
+        let location = values[0].location;
+
+        if values
+            .iter()
+            .all(|value| matches!(value.node, ExprKind::Constant { .. }))
+        {
+            // If every expression is a Constant, return a single Constant by concatenating the
+            // underlying strings.
+            Expr::new(
+                location,
+                ExprKind::Constant {
+                    value: Constant::Str(
+                        values
+                            .into_iter()
+                            .map(|value| {
+                                if let ExprKind::Constant {
+                                    value: Constant::Str(value),
+                                    ..
+                                } = value.node
+                                {
+                                    Ok(value)
+                                } else {
+                                    Err(LexicalError {
+                                        location,
+                                        error: LexicalErrorType::OtherError(
+                                            "Unexpected non-string constant.".to_string(),
+                                        ),
+                                    })
+                                }
+                            })
+                            .collect::<Result<Vec<String>, LexicalError>>()?
+                            .join(""),
+                    ),
+                    kind,
+                },
+            )
+        } else {
+            // Otherwise, we have at least one JoinedStr, so return a single JointedStr with as few
+            // Constants as possible by merging adjacent Constants.
+            let values: Vec<Expr> = values
+                .into_iter()
+                .map(|value| match value.node {
+                    ExprKind::JoinedStr { values } => Ok(values),
+                    ExprKind::Constant { .. } => Ok(vec![value]),
+                    _ => Err(LexicalError {
```

Comment:
Considering structuring this code such that I iterate once upfront to ensure that I only have these expression types (maybe even cast them to a custom Enum that's distinct from `ExprKind`?), so that I don't have to handle these mismatch errors multiple times.

Reply:
I suggested a way to use StringKind instead of Expr. That will be helpful if you want to go.